pkg/operator/ceph/object/objectstore.go

```diff
@@ -1010,3 +1017,31 @@ func errorOrIsNotFound(err error, msg string, args ...string) error {
 	}
 	return errors.Wrapf(err, msg, args)
 }
+
+// Update the Endpoints in Zone with new details
+func updateZoneEndpointList(zoneEndpointsList []string, oldEndpoint, newEndpoint string) []string {
+	appendNewEndpoint := false
+	if oldEndpoint != newEndpoint {
+		for i, endpoint := range zoneEndpointsList {
+			if oldEndpoint == endpoint {
+				zoneEndpointsList = append(zoneEndpointsList[:i], zoneEndpointsList[i+1:]...)
+				break
+			}
+		}
+		appendNewEndpoint = true
+	} else {
```

Comment:
@thotz I'm having a hard time following this logic. If the new endpoint is the same was the old endpoint isn't it already part of the zone's endpoint's list so we don't have to update anything?

It would be really helpful if you could go through some examples of the possibilities here and give a little background on the scenarios where a new endpoint would be used.

Reply:
I also agree. The code flow here is a bit hard to follow, but it seems largely correct.

Effectively, I think what needs to happen is this flow, right?
1. Rook determines whether the `spec.Endpoint` should be used or the Service endpoint should be used
2. From 1 above, pick the endpoint to use as `desiredEndpoint`, and the other as `undesiredEndpoint`
3. Update the endpoint list to remove the `undesiredEndpoint` and add the `desiredEndpoint`
4. Update the RGW's config

This function is effectively doing 2 and 3. I think it would be good to separate out those two pieces of logic so it's a bit easier to follow.

Reply:
Playing with this briefly, I am reminded that updating slices in-place in Go is often hard to read. Unless ultimate performance is the desire (which it generally isn't for the operator), I find it usually much easier to start with a new, empty slice and build it up with the desired elements.

Here's what I mean in the go playground: https://go.dev/play/p/fKo2QcqaHKW

Reply:
I think it would also be helpful to add a comment to the code to remind us in the future that we do want to leave other entries in the list because they may be endpoints from other CephObjectStores.

Reply:
Sure I will change it, always think so of modify existing than redefining functions which makes code bit complicated